extensions/amp-base-carousel/1.0/arrow.js

```diff
@@ -16,77 +16,86 @@
 
 import * as Preact from '../../../src/preact';
 import {useStyles} from './base-carousel.jss';
+import objstr from 'obj-str';
 
 /**
  * @param {!BaseCarouselDef.ArrowProps} props
  * @return {PreactDef.Renderable}
  */
 export function Arrow({
   advance,
+  as: Comp = DefaultArrow,
   by,
-  customArrow = <DefaultArrow by={by} />,
```

Comment:
Is the motivation here to allow the user to provide a Component, but for us to feed it extra props?

Reply:
Yes, the Component would be whatever arrow UI they want to use that is not our default grey circle with white arrow. The extra props will be state such as `disabled` and aria labels.

Reply:
This is the [relevant discussion](https://github.com/ampproject/amphtml/pull/33105#discussion_r588530945) for Lightbox's close button, wanted our approaches to be more unified going forward.

Reply:
Ok, that makes a lot of sense to me. And reviewing this file, it does seem like `as` should be the Component, and not a JSX node. Nice!

But, we **should not** use inline Components. Diffing works by using object identity, and if you were to have an inline `as: (props) => <button {...props} />`, you'll create a Component function (and a new object identity) every time. This will cause a full removal of the `<button>` DOM, create a new `<button>` DOM, and append that during every render. Components really need to be module-level constant values (or anything that's only evaluated once).

Reply:
Thanks for explaining. FMI, why doesn't using `Preact.cloneElement` yield new object identities?

Reply:
> @jridgewell  But we **should not** use inline Components.

Should we lint it? I wasn't aware of this problem myself.

Reply:
> why doesn't using Preact.cloneElement yield new object identities

It creates a new object identity for the VNode, but the underlying `Component` in `<Component />` can remain constant. VNode object identity is a minor performance optimization, and losing that because of `cloneElement(vnode, newProps)` won't hurts us much.

The issue I'm talking about here is in response to Alan's [comment](https://github.com/ampproject/amphtml/pull/33105#discussion_r588548661): "Can also set inline as a render prop: `<Lightbox closeButtonAs={(props) => <button {...props}>close</button>} />`". This crease a new `Component` identity during every render (and a new VNode identity, but that inconsequential in comparison).

When we look at the transformation of `<Foo attr="bar" />`, the VNode is roughly:

```js
function createElement(type: Component | string, props: Object): VNode {
  return { type, props };
}

const firstRender = createElement(Foo, { attr: "bar" });
const nextRender = createElement(Foo, { attr: "baz"});
```

Here, of course `firstRender.type === nextRender.type` (because they're both `Foo`). Alan's suggestion to inline a `Component` into the JSX (though an `as` prop) breaks this:

```js
function Foo(props) {
  const As = props.as;
  return <As />;
}

const firstRender = createElement(Foo, { as: () => <button /> });
const nextRender = createElement(Foo, { as() => <button /> });
```

If we render this `Foo` component, we return the VNode from its function body. So something like `createElement(props.as, {})`. The problem is `firstRender.props.as !== nextRender.props.as`. When React sees a VNode with a different `type` property, this means it's a wholly new Component Tree. React does not do diffing between transitioning to the wholly new Componet Tree. It unmounts the old Tree, and inserts the new tree. All state, DOM, everything is discarded.

> Should we lint it? I wasn't aware of this problem myself.

It's incredibly difficult to lint for. We could assume that any JSX attribute that ends in `as` is will be used as a VNode type, but as Caroline's code demonstrates, we can't hoist it any higher than the `updatePropsForRendering` method. We actually still have the bug, where any mutation by the publisher will cause a the buttons to be unmounted/inserted (because we'll get new Component identities once `updatePropsForRendering` is called.

I don't know how to solve this yet. It's possible we could copy from `oldProps` to `newProps` in this method, but the closed over `nextArrow`/`prevArrow` won't be updated that way.

Reply:
Can you clarify the next steps here? It sounds like we should be using named or memoized functions via `useCallback` and never anonymous ones. However `useCallback` can't be used outside of a functional component, so does not help us for the `updatePropsForRendering` piece. One thing I mentioned would be helpful as a followup in this PR description is to have `PreactBaseElement` support a prop configuration to give `Slot` instead of `<Slot>` -- would that help here?

Reply:
> would that help here?

That would allow us to eliminate these weird `cloneElement` components? Then, yes, that should be enough to fix this.

Reply:
Ok, in that case I'll do that PR first as a prereq to this one.